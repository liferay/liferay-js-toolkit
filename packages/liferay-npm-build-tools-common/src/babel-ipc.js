import {resolve} from 'path';

global._babel_ipc_ = global._babel_ipc_ || {};

/**
 * Get an IPC value for a given Babel plugin
 * @param  {Object} state babel plugin state argument
 * @param  {Object|function} defaultValue the default value or a factory to get it if value is not set
 * @return {Object} the IPC value
 */
export function get(state, defaultValue) {
	let defaultValueFactory = defaultValue;

	if (typeof defaultValue !== 'function') {
		defaultValueFactory = () => defaultValue;
	}

	if (
		!state ||
		!state.file ||
		!state.file.opts ||
		!state.file.opts.filenameRelative
	) {
		return defaultValueFactory();
	}

	const key = resolve(state.file.opts.filenameRelative);

	if (!global._babel_ipc_.hasOwnProperty(key)) {
		return defaultValueFactory();
	}

	return global._babel_ipc_[key];
}

/**
 * Set an IPC value for a given file path.
 * @param  {String} filePath the path of the file being processed
 * @param  {Object} value the IPC value to set
 * @return {void}
 */
export function set(filePath, value) {
	global._babel_ipc_[resolve(filePath)] = value;
}

/**
 * Clear an IPC value for a given file path.
 * @param  {String} filePath the path of the file being processed
 * @return {void}
 */
export function clear(filePath) {
	delete global._babel_ipc_[resolve(filePath)];
}
