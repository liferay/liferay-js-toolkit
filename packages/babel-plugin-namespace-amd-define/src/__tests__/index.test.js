import * as babel from 'babel-core';
import * as babelIpc from 'liferay-npm-build-tools-common/lib/babel-ipc';
import PluginLogger from 'liferay-npm-build-tools-common/lib/plugin-logger';
import plugin from '../index';

let logger;

beforeEach(() => {
	babelIpc.set(__filename, {
		log: (logger = new PluginLogger()),
	});
});

it('logs results correctly', () => {
	const source = `
	define([], function(){})
	if (typeof define === "function" && define.amd) {
		console.log('UMD!');
	}
	`;

	babel.transform(source, {
		filenameRelative: __filename,
		plugins: [plugin],
	});

	expect(logger.messages).toMatchSnapshot();
});

it('namespaces unqualified define calls', () => {
	const source = `
	define([], function(){})
	`;

	const {code} = babel.transform(source, {
		filenameRelative: __filename,
		plugins: [plugin],
	});

	expect(code).toMatchSnapshot();
});

it('does not namespace already qualified define calls', () => {
	const source = `
	Other.Namespace.define([], function(){})
	`;

	const {code} = babel.transform(source, {
		filenameRelative: __filename,
		plugins: [plugin],
	});

	expect(code).toMatchSnapshot();
});

it('namespaces references to global define identifier', () => {
	const source = `
	if (typeof define === "function" && define.amd) {
		console.log('UMD!');
	}
	`;

	const {code} = babel.transform(source, {
		filenameRelative: __filename,
		plugins: [plugin],
	});

	expect(code).toMatchSnapshot();
});

describe('does not namespace references to local define identifier', () => {
	it('when it is a variable', () => {
		const source = `
		let define = 'x';

		if(true) {
			if(true) {
				console.log(define);
			}
	    }
		`;

		const {code} = babel.transform(source, {
			filenameRelative: __filename,
			plugins: [plugin],
		});

		expect(code).toMatchSnapshot();
	});

	it('when it is a function parameter', () => {
		const source = `
		function fn(define) {
			console.log(define);
		};
		`;

		const {code} = babel.transform(source, {
			filenameRelative: __filename,
			plugins: [plugin],
		});

		expect(code).toMatchSnapshot();
	});

	it('when it is a function', () => {
		const source = `
		function define(x) {
			console.log(define);
		}
		`;

		const {code} = babel.transform(source, {
			filenameRelative: __filename,
			plugins: [plugin],
		});

		expect(code).toMatchSnapshot();
	});

	it('when it is an object field', () => {
		const source = `
			var a = {
				define: function(name, value) {
					return name;
				}
			}
			`;

		const {code} = babel.transform(source, {
			filenameRelative: __filename,
			plugins: [plugin],
		});

		expect(code).toMatchSnapshot();
	});
});
