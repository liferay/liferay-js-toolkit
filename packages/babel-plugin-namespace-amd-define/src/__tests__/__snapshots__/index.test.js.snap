// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly namespaces unqualified define calls 1`] = `
"
Liferay.Loader.define([], function () {});"
`;

exports[`does not namespaces qualified define calls 1`] = `
"
Other.Namespace.define([], function () {});"
`;
